iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/node/WALNode.java

```diff
@@ -255,91 +284,108 @@ public void run() {
           effectiveInfoRatio,
           costOfActiveMemTables,
           costOfFlushedMemTables);
+      WRITING_METRICS.recordWALNodeEffectiveInfoRatio(identifier, effectiveInfoRatio);
+    }
 
-      // try updating first valid version id by snapshotting or flushing memTable,
-      // then delete old .wal files again
-      if (!shouldSnapshotOrFlush()) {
+    private void summarizeExecuteResult() {
+      if (filesShouldDelete.length == 0) {
+        logger.info(
+            "wal node-{}:no wal file was found that should be deleted,current first valid version id is {}",
+            identifier,
+            firstValidVersionId);
         return;
       }
-      logger.debug(
-          "Effective information ratio {} (active memTables cost is {}, flushed memTables cost is {}) of wal node-{} is below wal min effective info ratio {}, some memTables will be snapshot or flushed.",
-          effectiveInfoRatio,
-          costOfActiveMemTables,
-          costOfFlushedMemTables,
-          identifier,
-          config.getWalMinEffectiveInfoRatio());
-      boolean isSuccess = snapshotOrFlushMemTable();
-      if (isSuccess && recursionTime < MAX_RECURSION_TIME) {
-        // wal is used to search, cannot optimize files deletion
-        if (safelyDeletedSearchIndex != DEFAULT_SAFELY_DELETED_SEARCH_INDEX) {
-          return;
+
+      buffer.subtractDiskUsage(deleteFileSize);
+      buffer.subtractFileNum(successfullyDeleted.size());
+      for (int i = 0; i < successfullyDeleted.size(); i++) {
+        long versionId = WALFileUtils.parseVersionId(successfullyDeleted.get(i));
+        Long memTableRamCostSum = walFileVersionId2MemTablesTotalCost.remove(versionId);
+        if (memTableRamCostSum != null) {
+          totalCostOfFlushedMemTables.addAndGet(-memTableRamCostSum);
         }
-        recursionTime++;
-        run();
+      }
+
+      if (!pinnedMemTableIds.isEmpty()
+          || fileIndexAfterFilterSafelyDeleteIndex < filesShouldDelete.length) {
+        StringBuilder summary =
+            new StringBuilder(
+                String.format(
+                    "wal node-%s delete outdated summary:the files that should be deleted is: [ %s ],successfully deleted file is: [ %s ],firstValidVersionId is %s .The following reasons influenced the result: %s",
+                    identifier,
+                    Arrays.stream(filesShouldDelete)
+                        .map(File::getName)
+                        .collect(Collectors.joining(",")),
+                    StringUtils.join(successfullyDeleted, ","),
+                    firstValidVersionId,
+                    System.getProperty("line.separator")));
+
+        if (!pinnedMemTableIds.isEmpty()) {
+          summary
+              .append("- MemTable has been flushed but pinned by PIPE,the MemTableId list is : ")
+              .append(StringUtils.join(pinnedMemTableIds, ","))
+              .append(".")
+              .append(System.getProperty("line.separator"));
+        }
```

Comment:
it seems that you will warn users when the PIPE pins some flushed memtables even if these operations aren't dangerous.